In [1]:
import sqlite3
import pandas as pd

In [2]:
df=pd.read_csv('Indian_Airlines_Cleaned.csv')

conn = sqlite3.connect(':memory:')
df.to_sql('Indian_Airlines',conn,index=False,if_exists='replace')

print("lenght of data :",len(df))

lenght of data : 300153


In [3]:
# 1. Aggregate average price by airline 
query1 = '''
select airline,avg(price) as average_price
from Indian_Airlines
group by airline
order by average_price desc
'''

result1 = pd.read_sql_query(query1,conn)
print('---'*10,'average price by airline : ','---'*10 ,'\n',result1)

------------------------------ average price by airline :  ------------------------------ 
      airline  average_price
0    Vistara   30396.536302
1  Air_India   23507.019112
2   SpiceJet    6179.278881
3   GO_FIRST    5652.007595
4     Indigo    5324.216303
5    AirAsia    4091.072742


In [4]:
# 2. Aggregate average price by route 
query2 = '''
select route,avg(price) as average_price
from Indian_Airlines
group by route
order by average_price desc
'''

result2 = pd.read_sql_query(query2,conn)
print('---'*10,'average price by route : ','---'*10 ,'\n',result2)


------------------------------ average price by route :  ------------------------------ 
                   route  average_price
0     Chennai-Bangalore   25081.850454
1       Kolkata-Chennai   23660.361040
2     Bangalore-Kolkata   23500.061229
3     Bangalore-Chennai   23321.850078
4      Mumbai-Bangalore   23147.873807
5      Bangalore-Mumbai   23128.618672
6        Mumbai-Chennai   22781.899112
7        Chennai-Mumbai   22765.849647
8     Kolkata-Bangalore   22744.808428
9       Chennai-Kolkata   22669.932407
10       Mumbai-Kolkata   22379.146723
11       Kolkata-Mumbai   22078.883579
12    Hyderabad-Chennai   21848.065989
13    Chennai-Hyderabad   21591.345404
14    Kolkata-Hyderabad   21500.011397
15  Hyderabad-Bangalore   21347.177998
16  Bangalore-Hyderabad   21226.121192
17     Mumbai-Hyderabad   21004.046705
18    Hyderabad-Kolkata   20823.893201
19        Delhi-Kolkata   20566.409418
20     Hyderabad-Mumbai   20080.865759
21        Kolkata-Delhi   19422.354559
22        Del

In [5]:
# 3a. Identifying the highest price flight
query3a='''
select airline, price
from Indian_Airlines
order by price desc
limit 1
'''

result3a = pd.read_sql_query(query3a,conn)
print('---'*10,'Top Price Airline : ','---'*10 ,'\n',result3a)

# 3b. Identifying the lowest rated flight
query3b='''
select airline,price
from Indian_Airlines
order by price asc 
limit 1 
'''

result3b = pd.read_sql_query(query3b,conn)
print('---'*10,'Lowest Price Airline : ','---'*10 ,'\n',result3b)

------------------------------ Top Price Airline :  ------------------------------ 
    airline   price
0  Vistara  123071
------------------------------ Lowest Price Airline :  ------------------------------ 
    airline  price
0  AirAsia   1105


In [6]:
# 4 What is the price variability (difference between max and min price) for each airline?
query4 ='''
select
  airline,
  max(price) - min(price) AS price_variability
from Indian_Airlines
group by airline
order by price_variability desc
'''
result4 = pd.read_sql_query(query4,conn)
print('---'*10,'price variability for each airline : ','---'*10 ,'\n',result4)

------------------------------ price variability for each airline :  ------------------------------ 
      airline  price_variability
0    Vistara             121357
1  Air_India              89444
2   SpiceJet              33052
3   GO_FIRST              31698
4     Indigo              30847
5    AirAsia              30812


In [7]:
# 5.Which airlines operate the most flights in the dataset?

query5 = '''
select airline,count(*) as total_flights
from Indian_Airlines
group by airline
order by total_flights desc
'''

result5 = pd.read_sql_query(query5,conn)
print('---'*10,'Airlines Operating the most flights : ','---'*10 ,'\n',result5)

------------------------------ Airlines Operating the most flights :  ------------------------------ 
      airline  total_flights
0    Vistara         127859
1  Air_India          80892
2     Indigo          43120
3   GO_FIRST          23173
4    AirAsia          16098
5   SpiceJet           9011


In [8]:
# 6.Which routes have the highest average flight prices?
query6a = '''
select route,avg(price) as average_price
from Indian_Airlines
group by route
order by average_price desc
limit 1
'''

result6a = pd.read_sql_query(query6a,conn)
print('---'*10,'highest average flight price by route : ','---'*10 ,'\n',result6a)

# 6b Which airlines offer the cheapest flights on a specific popular route
query6b = '''
select route,avg(price) as average_price
from Indian_Airlines
group by route
order by average_price asc
limit 1
'''
result6b = pd.read_sql_query(query6b,conn)
print('---'*10,'lowest average flight price by route : ','---'*10 ,'\n',result6b)

------------------------------ highest average flight price by route :  ------------------------------ 
                route  average_price
0  Chennai-Bangalore   25081.850454
------------------------------ lowest average flight price by route :  ------------------------------ 
              route  average_price
0  Hyderabad-Delhi   17243.945685


In [9]:
# 7.How do flight prices vary by time of day (departure hour)?
query7 = '''
SELECT 
  `departure_time`,
  avg(price) AS average_price
from Indian_Airlines
group by `departure_time`
order by `average_price` desc
'''

result7 = pd.read_sql_query(query7,conn)
print('---'*10,'flight price vary by time of day : ','---'*10 ,'\n',result7)


------------------------------ flight price vary by time of day :  ------------------------------ 
   departure_time  average_price
0          Night   23062.146808
1        Morning   21630.760254
2        Evening   21232.361894
3  Early_Morning   20370.676718
4      Afternoon   18179.203331
5     Late_Night    9295.299387


In [10]:
# 8.What are the top 5 cheapest and most expensive routes by average price?
query8a = '''
select route,avg(price) as average_price
from Indian_Airlines
group by route
order by average_price desc
limit 5
'''

result8a = pd.read_sql_query(query8a,conn)
print('---'*10,'Top 5 most expensive routes  : ','---'*10 ,'\n',result8a)

# 8b Which airlines offer the cheapest flights on a specific popular route
query8b = '''
select route,avg(price) as average_price
from Indian_Airlines
group by route
order by average_price asc
limit 5
'''
result8b = pd.read_sql_query(query8b,conn)
print('---'*10,' Top 5 cheapest routes  : ','---'*10 ,'\n',result8b)

------------------------------ Top 5 most expensive routes  :  ------------------------------ 
                route  average_price
0  Chennai-Bangalore   25081.850454
1    Kolkata-Chennai   23660.361040
2  Bangalore-Kolkata   23500.061229
3  Bangalore-Chennai   23321.850078
4   Mumbai-Bangalore   23147.873807
------------------------------  Top 5 cheapest routes  :  ------------------------------ 
              route  average_price
0  Hyderabad-Delhi   17243.945685
1  Delhi-Hyderabad   17347.288379
2  Bangalore-Delhi   17723.313972
3  Delhi-Bangalore   17880.216315
4     Mumbai-Delhi   18725.320008


In [11]:
#  9. Which airlines  and route have the highest average flight duration?

query9='''
select airline,route,avg(`total_duration (minutes)`) as average_duration
from Indian_Airlines
group by airline,route
order by average_duration desc
limit 5
'''

result9 = pd.read_sql_query(query9,conn)
print('---'*10,' the highest average flight duration : ','---'*10 ,'\n',result9)

------------------------------  the highest average flight duration :  ------------------------------ 
      airline                route  average_duration
0  Air_India      Chennai-Kolkata       1087.715891
1    Vistara      Kolkata-Chennai       1073.566071
2    Vistara      Chennai-Kolkata       1072.947779
3  Air_India    Chennai-Bangalore       1056.590354
4  Air_India  Bangalore-Hyderabad       1050.908887


In [12]:
# 10 How does average flight price vary by class?
query10 = '''
select class, avg(price) as average_price
from Indian_Airlines
group by class
order by average_price desc
'''

result10 = pd.read_sql_query(query10,conn)
print('---'*10,' Average price vary by class : ','---'*10 ,'\n',result10)

------------------------------  Average price vary by class :  ------------------------------ 
       class  average_price
0  Business   52540.081124
1   Economy    6572.342383


In [13]:
# 11 Which airlines offer the cheapest flights in each class?
query11 ='''
WITH RankedFlights AS (
    SELECT
        class,
        airline,
        MIN(price) AS cheapest_price,
        RANK() OVER (PARTITION BY class ORDER BY MIN(price)) AS rank
    FROM Indian_Airlines
    GROUP BY class, airline
)
SELECT class, airline, cheapest_price
FROM RankedFlights
WHERE rank = 1;
'''

result11 = pd.read_sql_query(query11,conn)
print('---'*10,'  cheapest flights in each class : ','---'*10 ,'\n',result11)

------------------------------   cheapest flights in each class :  ------------------------------ 
       class    airline  cheapest_price
0  Business  Air_India           12000
1   Economy    AirAsia            1105
2   Economy   GO_FIRST            1105
3   Economy     Indigo            1105


### Key Insights from SQL Analysis

- **Dataset Size**: 300,153 records.
- **Highest Average Price by Airline**: Vistara (~₹30,397).
- **Lowest Average Price by Airline**: AirAsia (~₹4,091).
- **Most Expensive Route (Avg)**: Chennai-Bangalore (~₹25,082).
- **Least Expensive Routes (Avg)**: Delhi-Hyderabad & Hyderabad-Delhi (~₹17,200).
- **Highest Single Ticket Price**: Vistara (₹123,071).
- **Lowest Single Ticket Price**: AirAsia (₹1,105).
- **Most Price Variability**: Vistara.
- **Lowest Price Variability**: AirAsia.